README
Pion Network Library: a C++ framework for building lightweight HTTP interfaces

Copyright (C) 2007 Atomic Labs, Inc.

Home page: http://www.atomiclabs.com/pion/net

See COPYING for licensing information.

Build and installation summary:

# ./configure [--with-boost=<BOOST_DIR>]
# make all
# make check	(this will build and run Pion's unit tests)
# make install	(this may require superuser/Administrator privileges)

Note that you only need to specify a "--with-boost=<BOOST_DIR>" parameter to
Pion's configure script if you installed Boost into an alternative directory
(other than /usr/local).

Detailed build instructions are available for all of the platforms supported
by the Pion within the "common/doc" subdirectory (README.platform).  Please
read these instructions first.

Starting with 0.4.2, Pion REQUIRES the Boost C++ libraries (version 1.34.1
or greater), and the Boost ASIO library (version 0.3.9 or greater). Please see
the README.boost file within the "common/doc" subdirectory for instructions on
how to download, build and install these libraries.

Boost ASIO additionally requires OpenSSL for SSL/TLS support and zlib for
compression.  You can download these libraries from the following locations:

http://www.openssl.org
http://www.zlib.net

For logging, Pion may be configured to:

a) use std::cout and std::cerr for logging (the default configuration)
b) use one of the following open source libraries: log4cxx, log4cpp or
   log4cplus (configure using one of --with-log4cxx, --with-log4cpp or
   --with-log4cplus, respectively)
c) disable logging entirely (configure --disable-logging)

You can download these logging libraries from the following locations:

http://logging.apache.org/log4cxx
http://log4cpp.sourceforge.net
http://log4cplus.sourceforge.net

Please note that not all of these libraries work on all platforms.  Please
see the README.platform files for more information.


ChangeLog
Version 3.0.21 (2011-01-25)
===========================

* TCPTimer.cpp was missing boost/bind.hpp include file (1005)
* TCPConnection - now allows for SSL state change (1024)
* HTTPMessage - added headers_only options for send() and receive() (1025)


Version 3.0.9 (2010-04-30)
==========================

* HTTPReader - add support for timeouts (900)


Version 3.0.5 (2010-04-02)
==========================

* TCPConnection - add accessors for ssl socket object (901)


Version 3.0.3 (2010-03-12)
==========================

* HTTPTypes::url_encode() crashes when input string has negative char (877)
* TCPConnection::write() was not passing error_code by reference (878)


Version 3.0.1 (2010-03-01)
==========================

* Added ability to dynamically remove resources from HTTPServers (834)
* Added support to HTTPParser for saving the raw HTTP headers (842)
* HTTPParser now uses a default status code of 0 instead of 200 (817)
* Removed build numbers from Unix library names (861)


Version 2.2.12 (2010-02-16)
===========================

* Made HTTP query parameter map case-insensitive (836)


Version 2.2.10 (2010-01-20)
===========================

* HTTPParser will now ignore attributes from Set-Cookie HTTP headers (808)
* Disabled default building of static libraries in Unix autoconf scripts (825)


Version 2.2.8 (2009-12-04)
==========================

* HTTP client library not sending payload content that started with NULL (777)
* Added support for newer Boost releases (through 1.41) and for gcc 4.4 (780)


Version 2.2.2 (2009-10-09)
==========================

* HTTPParser query parameter parsing failed on LF & CR characters (681)
* Improved autoconf detection for support of Boost.Lockfree library (692)
* Improved error handling in PluginManager when loading fails (736)
* Fixed "make install" on Unix platforms using autoconf (378)
* Pion's unit tests now use random port numbers (690)


Version 2.1.4 (2009-07-31)
==========================

* Switched default logging to use log4cplus 1.0.3 instead of log4cxx (274, 514)
* Updated and merged in the latest version of the boost.lockfree library (647)
* Added PionId class to replace dependency on the OSSP UUID library (515)
* The Unix build scripts now auto-detect everything correctly (591)
* Fixed copy methods for HTTPMessage, HTTPRequest & HTTPResponse (590)
* Added setMaxContentLength() method to HTTPServer (589)


Version 2.0.8 (2009-06-05)
==========================
* Fixed bug with "./configure --with-debug" failing on boost thread tests (598)


Version 2.0.6 (2009-05-26)
==========================
* HTTPParser exception thrown when Content-Length had leading whitespace (585)
* Miscellaneous improvements to Windows MSVC build projects (564)


Version 1.2.4 (2009-03-27)
==========================
* Removed dependencies on sqlite since it is now embedded into platform (475)
* HTTPParser dropping requests when TCP connection contained 304 responses (481)
* Updated PionLockedQueue so that it now always keeps track of the size (469)


Version 1.1.4 (2009-02-16)
==========================
* Build updated to Boost 1.36 & 1.37 (314, 270)
* Updates to support and use the latest Boost.Lockfree library (408, 409)
* Unit test cleanup and refactoring (416)


Version 1.0.10 (2009-01-22)
===========================
* Added support and unit tests for maximum HTTP payload content sizes (399)
* Updated HTTPParser to recover from missing packets and other errors (407)
* TCPServer can now bind to privileged ports when running as setuid (522)
* Fixed occasional crashes in pion-net unit tests on multi-core systems (528)


Version 1.0.8 (2008-12-17)
==========================
* Various improvements to how docs are built/cleaned (379)


Version 1.0.6 (2008-11-20)
==========================
* Pruning orphaned connections (346)
* Fixing auto generated INSTALL file (347)
* Autoconf fixes (347)


Version 1.0.4 (2008-10-16)
==========================
* Cookie collection now case sensitive (500)
* Added unit test and fixed bug when parsing query strings with "&&" (333)
* Added HEADER_X_FORWARDED_FOR to HTTPTypes (318)
* Added TCPConnection::connect() function that performs dns lookups (318)


Version 1.0.0 (2008-09-12)
==========================

* QueryService added, then removed, and placed into pion::platform
* Update clear() function to clear all members()
* Added test to see if base64 routines handle binary data in strings properly


Version 0.6.0 (2008-08-29)
==========================

* Fixed bug with the --disable-logging configure option breaking build (254)
* Updated HTTPParser::parseCookieHeader() to be usable for char[] as well
* Only HTML documentation is now build with Doxygen when using "make docs"


Version 0.5.8 (2008-07-18)
==========================

* Simplified HTTPParser class and eliminated some redundant code (240)
* HTTPTypes::url_encoded() was not properly encoding space characters (239)
* HTTPParser was not extracting POST query when charset is specified (243)
* Unable to compile HTTPAuth code with log4cplus support (249)
* TCPStream unit tests were occasionally failing (206)
* Added --with-boost-extension option to Unix configure script (217)
* Misc updates to get things building on gcc 4.2 and gcc 4.3 (241)


Version 0.5.6 (2008-05-02)
==========================

* Added support for user authentication using cookies and HTTP Basic-Auth
* Added support to HTTPServer to redirect/rewrite HTTP requests internally
* Updated HTTPRequest to track the original resource requested when redirected
* Refactored PionScheduler so that two scheduling policies are now provided
* Added PionLockedQueue and PionLockFreeQueue based on Michael & Scott's work
* Added "make docs" Unix build option that uses Doxygen to build all the docs
* Unix build scripts now support dynamic versions of the Boost.UnitTest libs
* Boost 1.35.0 or later is now REQUIRED (it includes ASIO); see README.boost


Version 0.5.4 (2008-03-07)
==========================


net/src/HTTPTypes.cpp
// ------------------------------------------------------------------
// pion-net: a C++ framework for building lightweight HTTP interfaces
// ------------------------------------------------------------------
// Copyright (C) 2007-2008 Atomic Labs, Inc.  (http://www.atomiclabs.com)
//
// Distributed under the Boost Software License, Version 1.0.
// See http://www.boost.org/LICENSE_1_0.txt
//

#include <boost/lexical_cast.hpp>
#include <boost/thread/mutex.hpp>
#include <pion/net/HTTPTypes.hpp>
#include <ctime>


namespace pion {		// begin namespace pion
namespace net {		// begin namespace net (Pion Network Library)


// generic strings used by HTTP
const std::string	HTTPTypes::STRING_EMPTY;
const std::string	HTTPTypes::STRING_CRLF("\x0D\x0A");
const std::string	HTTPTypes::STRING_HTTP_VERSION("HTTP/");
const std::string	HTTPTypes::HEADER_NAME_VALUE_DELIMITER(": ");

// common HTTP header names
const std::string	HTTPTypes::HEADER_HOST("Host");
const std::string	HTTPTypes::HEADER_COOKIE("Cookie");
const std::string	HTTPTypes::HEADER_SET_COOKIE("Set-Cookie");
const std::string	HTTPTypes::HEADER_CONNECTION("Connection");
const std::string	HTTPTypes::HEADER_CONTENT_TYPE("Content-Type");
const std::string	HTTPTypes::HEADER_CONTENT_LENGTH("Content-Length");
const std::string	HTTPTypes::HEADER_CONTENT_LOCATION("Content-Location");
const std::string	HTTPTypes::HEADER_CONTENT_ENCODING("Content-Encoding");
const std::string	HTTPTypes::HEADER_LAST_MODIFIED("Last-Modified");
const std::string	HTTPTypes::HEADER_IF_MODIFIED_SINCE("If-Modified-Since");
const std::string	HTTPTypes::HEADER_TRANSFER_ENCODING("Transfer-Encoding");
const std::string	HTTPTypes::HEADER_LOCATION("Location");
const std::string	HTTPTypes::HEADER_AUTHORIZATION("Authorization");
const std::string	HTTPTypes::HEADER_REFERER("Referer");
const std::string	HTTPTypes::HEADER_USER_AGENT("User-Agent");
const std::string	HTTPTypes::HEADER_X_FORWARDED_FOR("X-Forwarded-For");

// common HTTP content types
const std::string	HTTPTypes::CONTENT_TYPE_HTML("text/html");
const std::string	HTTPTypes::CONTENT_TYPE_TEXT("text/plain");
const std::string	HTTPTypes::CONTENT_TYPE_XML("text/xml");
const std::string	HTTPTypes::CONTENT_TYPE_URLENCODED("application/x-www-form-urlencoded");

// common HTTP request methods
const std::string	HTTPTypes::REQUEST_METHOD_HEAD("HEAD");
const std::string	HTTPTypes::REQUEST_METHOD_GET("GET");
const std::string	HTTPTypes::REQUEST_METHOD_PUT("PUT");
const std::string	HTTPTypes::REQUEST_METHOD_POST("POST");
const std::string	HTTPTypes::REQUEST_METHOD_DELETE("DELETE");

// common HTTP response messages
const std::string	HTTPTypes::RESPONSE_MESSAGE_OK("OK");
const std::string	HTTPTypes::RESPONSE_MESSAGE_CREATED("Created");
const std::string	HTTPTypes::RESPONSE_MESSAGE_NO_CONTENT("No Content");
const std::string	HTTPTypes::RESPONSE_MESSAGE_FOUND("Found");
const std::string	HTTPTypes::RESPONSE_MESSAGE_UNAUTHORIZED("Unauthorized");
const std::string	HTTPTypes::RESPONSE_MESSAGE_FORBIDDEN("Forbidden");
const std::string	HTTPTypes::RESPONSE_MESSAGE_NOT_FOUND("Not Found");
const std::string	HTTPTypes::RESPONSE_MESSAGE_METHOD_NOT_ALLOWED("Method Not Allowed");
const std::string	HTTPTypes::RESPONSE_MESSAGE_NOT_MODIFIED("Not Modified");
const std::string	HTTPTypes::RESPONSE_MESSAGE_BAD_REQUEST("Bad Request");
const std::string	HTTPTypes::RESPONSE_MESSAGE_SERVER_ERROR("Server Error");
const std::string	HTTPTypes::RESPONSE_MESSAGE_NOT_IMPLEMENTED("Not Implemented");
const std::string	HTTPTypes::RESPONSE_MESSAGE_CONTINUE("Continue");

// common HTTP response codes
const unsigned int	HTTPTypes::RESPONSE_CODE_OK = 200;
const unsigned int	HTTPTypes::RESPONSE_CODE_CREATED = 201;
const unsigned int	HTTPTypes::RESPONSE_CODE_NO_CONTENT = 204;
const unsigned int	HTTPTypes::RESPONSE_CODE_FOUND = 302;
const unsigned int	HTTPTypes::RESPONSE_CODE_UNAUTHORIZED = 401;
const unsigned int	HTTPTypes::RESPONSE_CODE_FORBIDDEN = 403;
const unsigned int	HTTPTypes::RESPONSE_CODE_NOT_FOUND = 404;
const unsigned int	HTTPTypes::RESPONSE_CODE_METHOD_NOT_ALLOWED = 405;
const unsigned int	HTTPTypes::RESPONSE_CODE_NOT_MODIFIED = 304;
const unsigned int	HTTPTypes::RESPONSE_CODE_BAD_REQUEST = 400;
const unsigned int	HTTPTypes::RESPONSE_CODE_SERVER_ERROR = 500;
const unsigned int	HTTPTypes::RESPONSE_CODE_NOT_IMPLEMENTED = 501;
const unsigned int	HTTPTypes::RESPONSE_CODE_CONTINUE = 100;


// static member functions
	
bool HTTPTypes::base64_decode(const std::string &input, std::string &output)
{
	static const char nop = -1; 
	static const char decoding_data[] = {
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop, 62, nop,nop,nop, 63,
		52, 53, 54,  55,  56, 57, 58, 59,  60, 61,nop,nop, nop,nop,nop,nop,
		nop, 0,  1,   2,   3,  4,  5,  6,   7,  8,  9, 10,  11, 12, 13, 14,
		15, 16, 17,  18,  19, 20, 21, 22,  23, 24, 25,nop, nop,nop,nop,nop,
		nop,26, 27,  28,  29, 30, 31, 32,  33, 34, 35, 36,  37, 38, 39, 40,
		41, 42, 43,  44,  45, 46, 47, 48,  49, 50, 51,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop,
		nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop, nop,nop,nop,nop
		};

	unsigned int input_length=input.size();
	const char * input_ptr = input.data();

	// allocate space for output string
	output.clear();
	output.reserve(((input_length+2)/3)*4);

	// for each 4-bytes sequence from the input, extract 4 6-bits sequences by droping first two bits
	// and regenerate into 3 8-bits sequence

	for (unsigned int i=0; i<input_length;i++) {
		char base64code0;
		char base64code1;
		char base64code2 = 0;   // initialized to 0 to suppress warnings
		char base64code3;

		base64code0 = decoding_data[static_cast<int>(input_ptr[i])];
		if(base64code0==nop)			// non base64 character
			return false;
		if(!(++i<input_length)) // we need at least two input bytes for first byte output
			return false;
		base64code1 = decoding_data[static_cast<int>(input_ptr[i])];
		if(base64code1==nop)			// non base64 character
			return false;

		output += ((base64code0 << 2) | ((base64code1 >> 4) & 0x3));

		if(++i<input_length) {
			char c = input_ptr[i];
			if(c =='=') { // padding , end of input
				BOOST_ASSERT( (base64code1 & 0x0f)==0);
				return true;
			}
			base64code2 = decoding_data[static_cast<int>(input_ptr[i])];
			if(base64code2==nop)			// non base64 character
				return false;

			output += ((base64code1 << 4) & 0xf0) | ((base64code2 >> 2) & 0x0f);
		}

		if(++i<input_length) {
			char c = input_ptr[i];
			if(c =='=') { // padding , end of input
				BOOST_ASSERT( (base64code2 & 0x03)==0);
				return true;
			}
			base64code3 = decoding_data[static_cast<int>(input_ptr[i])];
			if(base64code3==nop)			// non base64 character
				return false;

			output += (((base64code2 << 6) & 0xc0) | base64code3 );
		}

	}

	return true;
}

bool HTTPTypes::base64_encode(const std::string &input, std::string &output)
{
	static const char encoding_data[] = 
		"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

	unsigned int input_length=input.size();
	const char * input_ptr = input.data();

	// allocate space for output string
	output.clear();
	output.reserve(((input_length+2)/3)*4);

	// for each 3-bytes sequence from the input, extract 4 6-bits sequences and encode using 
	// encoding_data lookup table.
	// if input do not contains enough chars to complete 3-byte sequence,use pad char '=' 
	for (unsigned int i=0; i<input_length;i++) {
		int base64code0=0;
		int base64code1=0;
		int base64code2=0;
		int base64code3=0;

		base64code0 = (input_ptr[i] >> 2)  & 0x3f;	// 1-byte 6 bits
		output += encoding_data[base64code0];
		base64code1 = (input_ptr[i] << 4 ) & 0x3f;	// 1-byte 2 bits +

		if (++i < input_length) {
			base64code1 |= (input_ptr[i] >> 4) & 0x0f; // 2-byte 4 bits
			output += encoding_data[base64code1];
			base64code2 = (input_ptr[i] << 2) & 0x3f;  // 2-byte 4 bits + 

			if (++i < input_length) {
				base64code2 |= (input_ptr[i] >> 6) & 0x03; // 3-byte 2 bits
				base64code3  = input_ptr[i] & 0x3f;		  // 3-byte 6 bits
				output += encoding_data[base64code2];
				output += encoding_data[base64code3];
			} else {
				output += encoding_data[base64code2];
				output += '=';
			}
		} else {
			output += encoding_data[base64code1];
			output += '=';
			output += '=';
		}
	}

	return true;
}

std::string HTTPTypes::url_decode(const std::string& str)
{
	char decode_buf[3];
	std::string result;
	result.reserve(str.size());
	
	for (std::string::size_type pos = 0; pos < str.size(); ++pos) {
		switch(str[pos]) {
		case '+':
			// convert to space character
			result += ' ';
			break;
		case '%':
			// decode hexidecimal value
			if (pos + 2 < str.size()) {
				decode_buf[0] = str[++pos];
				decode_buf[1] = str[++pos];
				decode_buf[2] = '\0';
				result += static_cast<char>( strtol(decode_buf, 0, 16) );
			} else {
				// recover from error by not decoding character
				result += '%';
			}
			break;
		default:
			// character does not need to be escaped
			result += str[pos];
		}
	};
	
	return result;
}
	
std::string HTTPTypes::url_encode(const std::string& str)
{
	char encode_buf[4];
	std::string result;
	encode_buf[0] = '%';
	result.reserve(str.size());

	// character selection for this algorithm is based on the following url:
	// http://www.blooberry.com/indexdot/html/topics/urlencoding.htm
	
	for (std::string::size_type pos = 0; pos < str.size(); ++pos) {
		switch(str[pos]) {
		default:
			if (str[pos] > 32 && str[pos] < 127) {
				// character does not need to be escaped
				result += str[pos];
				break;
			}
			// else pass through to next case
		case ' ':	
		case '$': case '&': case '+': case ',': case '/': case ':':
		case ';': case '=': case '?': case '@': case '"': case '<':
		case '>': case '#': case '%': case '{': case '}': case '|':
		case '\\': case '^': case '~': case '[': case ']': case '`':
			// the character needs to be encoded
			sprintf(encode_buf+1, "%.2X", str[pos]);
			result += encode_buf;
			break;
		}
	};
	
	return result;
}	

std::string HTTPTypes::get_date_string(const time_t t)
{
	// use mutex since time functions are normally not thread-safe
	static boost::mutex	time_mutex;
	static const char *TIME_FORMAT = "%a, %d %b %Y %H:%M:%S GMT";
	static const unsigned int TIME_BUF_SIZE = 100;
	char time_buf[TIME_BUF_SIZE+1];

	boost::mutex::scoped_lock time_lock(time_mutex);
	if (strftime(time_buf, TIME_BUF_SIZE, TIME_FORMAT, gmtime(&t)) == 0)
		time_buf[0] = '\0';	// failed; resulting buffer is indeterminate
	time_lock.unlock();

	return std::string(time_buf);
}

std::string HTTPTypes::make_query_string(const QueryParams& query_params)
{
	std::string query_string;
	for (QueryParams::const_iterator i = query_params.begin(); i != query_params.end(); ++i) {
		if (i != query_params.begin())
			query_string += '&';
		query_string += url_encode(i->first);
		query_string += '=';
		query_string += url_encode(i->second);
	}
	return query_string;
}

std::string HTTPTypes::make_set_cookie_header(const std::string& name,
											  const std::string& value,
											  const std::string& path,
											  const bool has_max_age,
											  const unsigned long max_age)
{
	std::string set_cookie_header(name);
	set_cookie_header += "=\"";
	set_cookie_header += value;
	set_cookie_header += "\"; Version=\"1\"";
	if (! path.empty()) {
		set_cookie_header += "; Path=\"";
		set_cookie_header += path;
		set_cookie_header += '\"';
	}
	if (has_max_age) {
		set_cookie_header += "; Max-Age=\"";
		set_cookie_header += boost::lexical_cast<std::string>(max_age);
		set_cookie_header += '\"';
	}
	return set_cookie_header;
}

	
}	// end namespace net
}	// end namespace pion

